taskcat/_cli_modules/test.py

```diff
@@ -133,90 +123,33 @@ def run(  # noqa: C901
-            config.config.project.name,
-            tests,
-            shorten_stack_name=config.config.project.shorten_stack_name,
+        :param skip_upload: Use templates in an existing cloudformation bucket.
+        """  # noqa: B950
+
+        test = CFNTest.from_file(
+            project_root=project_root,
+            input_file=input_file,
+            regions=regions,
+            enable_sig_v2=enable_sig_v2,
         )
-        test_definition.create_stacks()
+
+        # This code is temporary and should be removed once its easier
+        # to create a config object
+        frame = inspect.currentframe()
+
+        if frame is not None:
+            args, _, _, values = inspect.getargvalues(frame)
+
+            for i in args:
+                if hasattr(test, i):
+                    setattr(test, i, values[i])
+
         terminal_printer = TerminalPrinter(minimalist=minimal_output)
-        # 5. wait for completion
-        terminal_printer.report_test_progress(stacker=test_definition)
-        status = test_definition.status()
-        # 6. create report
-        report_path = Path(output_directory).resolve()
-        report_path.mkdir(exist_ok=True)
-        cfn_logs = _CfnLogTools()
-        cfn_logs.createcfnlogs(test_definition, report_path)
-        ReportBuilder(test_definition, report_path / "index.html").generate_report()
-        # 7. delete stacks
-        if no_delete:
-            LOG.info("Skipping delete due to cli argument")
-        elif keep_failed:
-            if len(status["COMPLETE"]) > 0:
-                LOG.info("deleting successful stacks")
-                test_definition.delete_stacks({"status": "CREATE_COMPLETE"})
-                if not dont_wait_for_delete:
-                    terminal_printer.report_test_progress(stacker=test_definition)
-        else:
-            test_definition.delete_stacks()
-            if not dont_wait_for_delete:
-                terminal_printer.report_test_progress(stacker=test_definition)
-        # TODO: summarise stack statusses (did they complete/delete ok) and print any
-        #  error events
-        # 8. delete buckets
 
-        if not no_delete or (keep_failed is True and len(status["FAILED"]) == 0):
-            deleted: ListType[str] = []
-            for test in buckets.values():
-                for bucket in test.values():
-                    if (bucket.name not in deleted) and not bucket.regional_buckets:
-                        bucket.delete(delete_objects=True)
-                        deleted.append(bucket.name)
-        # 9. raise if something failed
-        # - grabbing the status again to ensure everything deleted OK.
-        status = test_definition.status()
-        if len(status["FAILED"]) > 0:
-            raise TaskCatException(
-                f'One or more stacks failed tests: {status["FAILED"]}'
-            )
+        test.printer = terminal_printer
+
+        with test:
+            test.report(output_directory)
```

Comment:
So this is the only change in behavior of `taskcat test run`. In the past if some kind of exception was thrown then taskcat would exit. This would lead to stacks and s3 buckets left in the account. Now if an exception occurs the test will still call `clean_up` which I belive will delete the stacks and buckets.